src/client/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs

```diff
@@ -86,18 +87,34 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
             // No AT in the cache or AT needs to be refreshed
             try
             {
-                using (logger.LogBlockDuration("[OBO request] Fetching OBO token from ESTS"))
-                {
-                    var result = await FetchNewAccessTokenAsync(cancellationToken).ConfigureAwait(false);
-                    return result;
-                }
+                return await RefreshRtOrFetchNewAccessTokenAsync(cancellationToken).ConfigureAwait(false);
             }
             catch (MsalServiceException e)
             {
                 return HandleTokenRefreshError(e, msalAccessTokenItem);
             }
         }
 
+        private async Task<AuthenticationResult> RefreshRtOrFetchNewAccessTokenAsync(CancellationToken cancellationToken)
+        {
+            // Look for a refresh token
+            MsalRefreshTokenCacheItem appRefreshToken = await CacheManager.FindRefreshTokenAsync().ConfigureAwait(false);
+
+            // If a refresh token is not found, fetch a new access token
+            if (appRefreshToken == null)
```

Comment:
nit: code is a bit more readable if you handle the case where you get a refresh token back first 

`if (appRefreshToken!=null) { ... } `